### ISSUE TITLE
Backstage Carbon Plugin: Add cache config on app-config.yaml 

### ISSUE THREAD

**Issue by jonathan-arenas-edo** [cloud-carbon-footprint/ccf-backstage-plugin#75]
Hi team,

I am working with the latest carbon footprint version:
```
@cloud-carbon-footprint/backstage-plugin-frontend@0.0.8
@cloud-carbon-footprint/backstage-plugin-backend@0.1.1
@cloud-carbon-footprint/common": "^1.11.0
```
Backstage version:
```
Version:
OS: Darwin 22.4.0 - darwin/arm64
node: v16.19.0
yarn: 1.22.19
cli: 0.22.7 (installed)
backstage: 1.14.1
```

Configured backstage:
```
cloudCarbonFootprint:
      client:
        dateRangeType: years
        dateRangeValue: 1
        groupBy: week
      gcp:
        useBillingData: true
        billingProjectId: xxxxxxxxx
        billingProjectName: xxxxxxxxx
        bigQueryTable: billing.gcp_billing_export_v1_xxxxxxxxx
```
Is possible add the cache config on app-config.yaml? I want to use a GCS solution for cache storage:
For mongodb is possible, but for GCP is possible too?
https://github.com/cloud-carbon-footprint/ccf-backstage-plugin/tree/trunk/plugins/backend#caching

Kind regards.

**Comment by 4upz** [comment]
Hi @jarenasla-edo, thanks for opening this issue! Are you able to provide a screenshot to help us understand what's going on?

To make sure I understand correctly, when loading the CCF page you see data for the breakdown charts and other components but not the usage graph? And recommendations indefinitely load?

**Comment by 4upz** [comment]
Sorry I believe I replied before you edited the issue. In regards to configuring a cache solution, this should already be accessible in the config. You can refer to this template to see where this would go: https://github.com/cloud-carbon-footprint/ccf-backstage-plugin/blob/trunk/plugins/backend/config.d.ts#L236

However, feel free to let us know if you have any trouble with configuring GCS with that variable. We're happy to troubleshoot if it gives you issues.

**Comment by jonathan-arenas-edo** [comment]
Hi @4upz , thanks for your comments.

I tested the GCS option but the system don't create a local cache file or without GCS, in local the file is not exists.
My problem is that the plugin need 30s to load data, and every time need anothers 30s for this reason I need implement the cache.

```
cloudCarbonFootprint:
      client:
        dateRangeType: months
        dateRangeValue: 6
        groupBy: month
      gcp:
        useBillingData: true
        billingProjectId: xxxxxxx
        billingProjectName: xxxxxxx
        bigQueryTable: billing.gcp_billing_export_v1_xxxxxxx
        gcsCacheBucketName: xxxxxxx-cache 
      optional:
        cacheMode: GCS
```
I watched on my events the next line:
`2023-06-22T16:23:17.739Z [Cache] info: Ignoring cache... `

Do you know because the system ignore the cache?
The option disableCache by default is false, is correct because I want enabled the cache.
https://github.com/cloud-carbon-footprint/ccf-backstage-plugin/blob/trunk/plugins/frontend/config.d.ts

Kind regards.

**Comment by 4upz** [comment]
@jarenasla-edo I appreciate the context, the cache is definitely helpful for reducing the load times. The app should not be ignoring cache by default. As a sanity check, have you tried explicitly setting `disableCache` to false in you config?

Additionally, would you mind providing an example of the request url that the client is making? You should be able to see it in your networking tab using your browser's debug tools.

Another possibility is that you may have had cache configured to previously be disable and have not restarted your app since changing the config.

**Comment by jonathan-arenas-edo** [comment]
Hi @4upz ,

Yes I reviews my config and the disableCache is false, but the request ignore the cache:
`cloud-carbon-footprint/footprint?start=2023-05-23&end=2023-06-23&ignoreCache=true&groupBy=month`

I restarted the app on each config change. (I use argocd to deploy new version) There are another config part on I need configure the cache?

Config file:

```
cloudCarbonFootprint:
      client:
        dateRangeType: months
        dateRangeValue: 6
        groupBy: month
        disableCache: false
      gcp:
        useBillingData: true
        billingProjectId: xxxxxxx
        billingProjectName: xxxxxxx
        bigQueryTable: billing.gcp_billing_export_v1_xxxxxxx
        gcsCacheBucketName: xxxxxxx-cache 
      optional:
        cacheMode: GCS
```

But the cache is not activate.
Can you tell me why this happens?

![Captura de pantalla 2023-06-23 a las 8 27 49](https://github.com/cloud-carbon-footprint/ccf-backstage-plugin/assets/118807917/58968704-7cac-408e-8677-fe1600d23783)

**Comment by jonathan-arenas-edo** [comment]
Hi @4upz ,

On the other hand I have detected that the on the 3 call to footprint plugin fail, the plugin return an error 503 (undefined:1)
The plugin is working with backstage version 1.11: https://github.com/cloud-carbon-footprint/ccf-backstage-plugin/blob/trunk/backstage.json

Perhaps are there are any conflict with my version 1.14.1 (not the latest)?

Maybe is possible to schedule a meeting in order to improve and review the plugin together? 
I detected some anomalous behavior for example on RECOMMENDATIONS. (I will commented after solving the previous errors)

Thank you for your effort and help.

**Comment by 4upz** [comment]
@jarenasla-edo it seems that the initial request is including the cache but it is making a second request where it ignores it. 

I don't think the latest version of Backstage would cause this type of behavior. However, we can investigate on our side to see confirm whether this is the case. Happy to also meet if you want to [email me](mailto:arik.smith@thoughtworks.com) your availability for next week as well as your timezone.

**Comment by jonathan-arenas-edo** [comment]
Hi @4upz 

Thank you for your help! I will send you an email to schedule a meeting.

Kind regards :)

**Comment by camcash17** [comment]
Just made some updates to the plugin. Looks like the main issue was that it was no longer compatible with the major changes we recently did with the CCF client package. After updating the imports/exports, there are still a few issues to address before this is fully resolved:

Additional Bugs:
- There is currently a react version mismatch causing a useState error when loading the CCF route. I believe this is being caused by the CCF client react version and the backstage create-app react version, so we may need to update peer-dependencies. Current workaround until solved is to delete the react dependency from the CCF client node_modules and refresh
- I believe with the new changes, the recommendations table UI was altered and now it is difficult to see. There is also a bug where if you navigate away from the recommendations page and then return, the recommendation data is not displayed.
- After going back and forth between backstage routes, you will end up getting the following error:
`SyntaxError: Unexpected token ] in JSON at position 0` which we work-around by deleting and re-creating the cache file. we may want to understand why this continues to happen.

Additional Notes:
- We still have additional plugin config to add (ie start/end dates in the client) along with logic to handle.
- There are a few console warnings we should look into.
- We are not currently passing through API error into the useFootprint hook like we do in CCF client. we may want to consider adding this.

**Comment by jonathan-arenas-edo** [comment]
Hi @camcash17 

Thank you for your feedback. 
I'm going to update the plugin and I will pending to new versions in order to fix the problems commented.

Kind regards.

**Comment by jonathan-arenas-edo** [comment]
Hi @camcash17 ,

I updated to latest version, but the plugin is failing....

```
Error
TypeError

Message
Cannot read properties of null (reading 'useState')

Stack Trace
TypeError: Cannot read properties of null (reading 'useState')
    at useState (webpack-internal:///../../node_modules/@cloud-carbon-footprint/client/node_modules/react/cjs/react.development.js:1623:21)
    at useRemoteFootprintService (webpack-internal:///../../node_modules/@cloud-carbon-footprint/client/dist/utils/hooks/FootprintServiceHook.js:12:50)
    at useFootprintData (webpack-internal:///../../node_modules/@cloud-carbon-footprint/client/dist/utils/hooks/FootprintDataHook.js:10:73)
    at PluginTabs (webpack-internal:///../../node_modules/@cloud-carbon-footprint/backstage-plugin-frontend/dist/esm/Plugin-0c8d69e6.esm.js:115:101)
    at renderWithHooks (webpack-internal:///../../node_modules/react-dom/cjs/react-dom.development.js:14985:18)
    at mountIndeterminateComponent (webpack-internal:///../../node_modules/react-dom/cjs/react-dom.development.js:17811:13)
    at beginWork (webpack-internal:///../../node_modules/react-dom/cjs/react-dom.development.js:19049:16)
    at HTMLUnknownElement.callCallback (webpack-internal:///../../node_modules/react-dom/cjs/react-dom.development.js:3945:14)
    at Object.invokeGuardedCallbackDev (webpack-internal:///../../node_modules/react-dom/cjs/react-dom.development.js:3994:16)
    at invokeGuardedCallback (webpack-internal:///../../node_modules/react-dom/cjs/react-dom.development.js:4056:31)
```

![Captura de pantalla 2023-07-04 a las 9 40 11](https://github.com/cloud-carbon-footprint/ccf-backstage-plugin/assets/118807917/0baeb302-7864-4084-abe4-b1147b51201c)

**Comment by 4upz** [comment]
Hi @jarenasla-edo this fix is still ongoing. @camcash17 comments were meant to only be progress updates. 

Apologies for the confusion! We will let you know when it's released and ready to test.

**Comment by jonathan-arenas-edo** [comment]
Great @4upz ! Thank you :)

**Comment by ccasher** [comment]
In the v4 release of `@cloud-carbon-footprint/client` package, we introduced React 18. We did not update the CCF Client package in the Backstage Plugin until [this recent commit](https://github.com/cloud-carbon-footprint/ccf-backstage-plugin/commit/52150825508348c01831f0df712d51872faef89d#diff-ec627838686016983ce2b3072939b168c8679ec3721ae88bcdd630fffaca57b7), which is when we began to see compatibility issues.

We have since updated some logic in the plug-in, but have determined that in order to avoid the `useState error` when hitting the CCF route, you will need to add the following to the resolutions object in the root package.json of your Backstage instance:

```diff
  "resolutions": {
    "@types/react": "^17",
    "@types/react-dom": "^17",
+   "react": "^17",
    }
```
After a fresh `yarn install`, the plug-in should work properly. We will plan to create additional issues for the outstanding tasks mentioned in my [last comment](https://github.com/cloud-carbon-footprint/ccf-backstage-plugin/issues/75#issuecomment-1615100381)

**Comment by 4upz** [comment]
@ccasher I think this is a good workaround until [official support for React 18](https://github.com/backstage/backstage/milestone/41) is added to Backstage. I think it's a low-lift solution for those setting up a new CCF plugin.

@jarenasla-edo Could you verify that his recommendation works for you?

**Comment by coreyetsy** [comment]
Hello! I had a [similar issue](https://groups.google.com/g/cloud-carbon-footprint/c/9f4PN84TE3U) and tried this

 ```
 "resolutions": {
    "@types/react": "^17",
    "@types/react-dom": "^17",
    "react": "^17"
  },
```

unfortunately I'm still getting the same problem:

```
Cannot read properties of null (reading 'useState')
Call Stack
 useState
  node_modules/@cloud-carbon-footprint/client/node_modules/react/cjs/react.development.js:1623:21
 useRemoteFootprintService
  node_modules/@cloud-carbon-footprint/client/dist/utils/hooks/FootprintServiceHook.js:12:50
 useFootprintData
  node_modules/@cloud-carbon-footprint/client/dist/utils/hooks/FootprintDataHook.js:10:73
 PluginTabs
  node_modules/@cloud-carbon-footprint/backstage-plugin-frontend/dist/esm/Plugin-0c8d69e6.esm.js:115:101
 renderWithHooks
  node_modules/react-dom/cjs/react-dom.development.js:14985:18
 mountIndeterminateComponent
  node_modules/react-dom/cjs/react-dom.development.js:17811:13
 beginWork
  node_modules/react-dom/cjs/react-dom.development.js:19049:16
 HTMLUnknownElement.callCallback
  node_modules/react-dom/cjs/react-dom.development.js:3945:14
 Object.invokeGuardedCallbackDev
  node_modules/react-dom/cjs/react-dom.development.js:3994:16
 invokeGuardedCallback
  node_modules/react-dom/cjs/react-dom.development.js:4056:31
```

Any other suggestions are welcome! Thanks...
Corey

**Comment by ccasher** [comment]
Hi @coreyetsy, could you please confirm you added the resolutions to the root package.json of your backstage application?

Also, could you run the following as a sanity check:
```
rm yarn.lock
rm -rf node_modules
yarn install
```

**Comment by ccasher** [comment]
Closing this as the conversation has gone stale.

**Comment by zentala** [comment]
The same error occurred when developing a custom plugin with the latest Backstage version (v1.30, which has already fully migrated to React 18).

Thanks to your suggestions, I found the source of the issues. It seems that when creating a new plugin with `backstage-cli`, it still includes React v17 as a dependency. 

I solved this problem by replacing the React version definition in the plugin's `package.json` from `^16.13.1 || ^17.0.0 || ^18.0.0` to `^18.0.0`. Later, as suggested here, I had to remove the `node_modules` directory and `yarn.lock` file, then install the dependencies again.

After these changes, my code using `useState` started working like a charm, resolving this the React-related issue!

Hope it may help someone in the future!
